Feat : order system 구현

YooJisu826 opened the pull request:
### 요약
Customer, Menu, Kiosk, HomeworkMain 클래스로 구성되어 있고, 문제의 모든 기능을 구현하였습니다.
추가로 잔고를 충전하는 기능도 구현하였습니다.

### 어려웠던 점
main 함수 내용을 더 객체지향적으로 바꿔보고 싶었는데(start 함수를 만들어 초기 입력을 받는 등..), scanner가 문제가 되어 일단 다시 되돌렸습니다.

### 코드 리뷰 받고 싶은 점
클래스를 많이 나누었는데 함수들이 올바른 위치에 작성된 것이 맞는 지, 혹시나 더 복잡해진 점이 있는지 궁금합니다.
그리고 객체지향적 관점에서 어떤 점을 보완하면 좋을 지 궁금합니다.

YooJisu826 committed:
Feat : order system 구현

Customer, Menu, Kiosk, HomeworkMain 클래스로 구성

JuneYoungDo commented on src/main/java/com/landvibe/homework/Customer.java:
```diff
@@ -0,0 +1,42 @@
+    String name;
+    int money;
+    //고객이 주문한 주문 번호를 저장해놓는 정수형 벡터
+    Vector<Integer> allOrder = new Vector<Integer>();
```
자바에서는 다양한 종류의 배열 형태가 있어
c++에서 벡터가 굉장히 유용해서 많이 사용해 익숙하겠지만 자바에서는 차이를 알아보고 사용하는게 좋다고 생각해
아래 링크를 보면 알겠지만 자바에서는 vector의 사용을 지양하는 편이야!

[http://www.tcpschool.com/java/java_collectionFramework_list](url)

JuneYoungDo commented on src/main/java/com/landvibe/homework/Customer.java:
```diff
@@ -0,0 +1,42 @@
+
+    void setAllOrder(int i) {
+        allOrder.add(i);
+    }
```
setAllOrder이라고 해서 모든 메뉴를 저장해 주는 건가? 라고 생각했었는데 내용을 보면 하나의 주문을 추가해 주는 함수더라구
의미는 비슷하지만, 좀 더 직관적인 네이밍이 있으면 협업하는 사람들 입장에서 더 보기 좋겠지??
만약 setAllOrder이라는 함수가 굉장히 길어졌을 경우 함수의 이름만 보고 잘못 이해하는 경우가 발생할 수 있다고 생각해

JuneYoungDo commented on src/main/java/com/landvibe/homework/Menu.java:
```diff
@@ -0,0 +1,22 @@
+public class Menu {
+    // 메뉴 이름, 메뉴 가격을 저장해놓는 배열들 (index 1부터 시작)
+    String[] menuName = {"이름", "짜장면" , "간짜장", "짬뽕", "짬뽕밥", "차돌짬뽕"};
+    int[] menuPrice = {0, 6000, 6500, 7000, 7000, 9000};
```
자바에서 리스트를 표현하는 많은 방법이 있는데, array와 arrayList의 차이는 알고 쓰는게 좋을거같아
[https://dev-coco.tistory.com/15](url)

추후를 생각했을 때는, 가변적인 arrayList가 더 효율적이지 않을까??

또 메뉴를 구성하는 방법에 대해서 잘 구현해준거 같아
한가지 더 추가해 보자면 여기서 메뉴라는 객체를 따로 하나 만들어주고, 메뉴의 리스트를 하나 만들어주게 되면 더 객체지향적인 코드가 되지 않을까?? 또 배열의 index가 아니라 메뉴마다 고유한 자체 번호를 부여한다면 이후 사용성이나 확장성을 고려했을 때도 더 효과적으로 쓰일 수 있을거 같아

KeonHee replied:
menuName와 menuPrice가 index로 묶이는 구조가 어색해보입니다. 실제로 위에 이 상황을 설명할 주석이 없으면 실수할 여지도 있구요.

Menu 클래스에 `String name`, `int price`만 두고 Menu를 리스트로 관리하면 어떨까요?

```java
List<Mene> menus = ArrayList();
menus.add(new Menu("짜장면", 6000);
...
```

JuneYoungDo commented on src/main/java/com/landvibe/homework/Menu.java:
```diff
@@ -0,0 +1,22 @@
+
+    int getMenuPrice(int i) { return menuPrice[i]; };
+
+    int getMenuLength() { return menuName.length; };
```
혹시 이 함수는 어디에 쓰이는 것인지 알 수 있을까요..❓

kang-jisu commented on src/main/java/com/landvibe/homework/Customer.java:
```diff
@@ -0,0 +1,42 @@
+        this.money = money;
+    }
+
+    String getName() { return name; };
+
+    int getMoney() { return money; };
+
```
클래스의 인스턴스를 생성해서 필드를 그냥 가져다 사용하지않고 getter, setter를 만드는 이유가 뭘까 !
클래스 내부 필드를 숨기고 캡슐화를 하려면 어떻게 자바의 접근제어자를 활용할 수 있는지 한번 알아보고 적용해보면 좋을것같아!!
- [접근제어자](http://www.tcpschool.com/java/java_modifier_accessModifier)

kang-jisu commented on src/main/java/com/landvibe/homework/HomeworkMain.java:
```diff
@@ -1,7 +1,74 @@
+        int money = scanner.nextInt();
+        String line = scanner.nextLine(); // 남아있는 \n를 없애주기 위함
+
+        // customer, kiosk 객체 생성
+        Customer customer = new Customer(name, money);
+        Kiosk kiosk = new Kiosk(customer);
```
👍

kang-jisu commented on src/main/java/com/landvibe/homework/Kiosk.java:
```diff
@@ -0,0 +1,76 @@
+import com.landvibe.homework.Customer;
+import com.landvibe.homework.Menu;
+
+public class Kiosk {
+    Customer customer;
+    Menu menu = new Menu();
+
+    // 생성자, customer의 name과 money를 받아 초기화
+    Kiosk(Customer customer) {
+        this.customer = customer;
+    }
```
Customer는 생성자에서 인자로 받아 초기화하고, menu는 바로 초기화하고있는데
그럼 Menu는 어떤 Kiosk 인스턴스가 생성되어도 항상 같은 형태(?)로 인스턴스가 생성되겠지 ! 
Kiosk객체가 100개가 생기는데 그 안에 항상 같은 Menu가 100개 생기면 메모리적으로 낭비이지않을까 !! 
final, static과 같은 키워드 의미를 한번 알아보고 어떻게사용하면 더 효율적인 객체 생성을 할 수 있을지 한번 고민해보면 좋을거같아 !!

kang-jisu reviewed (commented):
역할에 따라서 클래스도 분리하고 신경써서 구현 잘 해본것같아!!
몇개 코멘트 남긴것들도 생각해봐주면 좋을듯~ !!

YooJisu826 committed:
Update : Menu를 다루는 구조 수정

Vector를 List로 변경, allOrder 필드명 변경, 접근 제어자 사용

YooJisu826 commented:
코멘트 남겨주셔서 감사합니다!! 코드 일부분 수정해서 커밋했습니다!

kang-jisu commented on src/main/java/com/landvibe/homework/Kiosk.java:
```diff
@@ -59,16 +75,21 @@ void orderMenu(String order) {
+
                 System.out.println("이름 : " + customer.getName());
-                System.out.println("메뉴 : " + menu.getMenuName(allOrder.get(i)) + ", 가격 : " + menu.getMenuPrice(allOrder.get(i)) + "원");
+                System.out.println("메뉴 : " + orderedMenu.getMenuName() + ", 가격 : " + orderedMenu.getMenuPrice() + "원");
```
```java
// Customer
private List<Integer> orderHistory = new ArrayList<>();
```
굳이 orderHistory 에서 Integer를 저장하고 그걸로 다시 menuList에서 get해서 가져올필요가 있을까!?
그냥 `List<Menu>`로 저장하는 방법을 생각해보면 좋을 것 같아!!

kang-jisu commented on src/main/java/com/landvibe/homework/Menu.java:
```diff
@@ -1,22 +1,15 @@
-    int[] menuPrice = {0, 6000, 6500, 7000, 7000, 9000};
-
-    String getMenuName(int i) { return menuName[i]; };
+    private String menuName;
+    private int menuPrice;
```
https://github.com/LandvibeDev/2022-Backend-SummerCoding/pull/6#discussion_r922822677
이 링크 한번 확인해줘~

kang-jisu commented on src/main/java/com/landvibe/homework/Kiosk.java:
```diff
@@ -24,27 +35,32 @@ void printOption() {
-    void orderMenu(String order) {
+    public void orderMenu(String order) {
         //해당 메뉴가 있다면 true로 변경! 해당 메뉴가 없다면 "해당 메뉴가 없습니다" 출력
         boolean isExist = false;
```
isExist가 해당 메뉴가 있는지 확인하는 부분인거같은데, 아래 if else구문에 상관없이
```java
isExist = true;
break;
```
처리를 해주는게 있네! 
두개를 그냥 조건문에 상관없이 한곳에서만 처리해줘도 좋을 것 같당

kang-jisu commented on src/main/java/com/landvibe/homework/Kiosk.java:
```diff
@@ -0,0 +1,97 @@
+        this.customer = customer;
+
+        // Menu 리스트인 menuList를 기본 메뉴들로 초기화!
+        menuList.add(new Menu( "메뉴이름", 0));
```
개인적으로 출력을 위해서 menuList인데 메뉴이름,0 이라는 의미없는 값을 넣는게 맞는건지는 잘 모르겠어 ㅠㅠ 
이것때문에 계속 menuList를 쓸때 0번 인덱스를 제외하고 처리해줘야하는부분도 있고..!

kang-jisu commented on src/main/java/com/landvibe/homework/Kiosk.java:
```diff
@@ -0,0 +1,97 @@
+        List<Integer> orderHistory = customer.getOrderHistory();
+
+        // 주문 내역 리스트가 비어있다면
+        if(orderHistory.isEmpty()) {
+            System.out.println("주문 내역이 존재하지 않습니다.");
+        }
+        // 비어있지 않다면
+        else {
+            for (int i = 0; i < orderHistory.size(); i++) {
+                int menuIndex = orderHistory.get(i); // 주문 번호 가져오기
+                Menu orderedMenu = menuList.get(menuIndex); // 주문 번호를 이용하여 해당 Menu 객체 생성
+
+                System.out.println("이름 : " + customer.getName());
+                System.out.println("메뉴 : " + orderedMenu.getMenuName() + ", 가격 : " + orderedMenu.getMenuPrice() + "원");
+            }
+        }
```
if조건문에서 할게 끝났다면 return을 해버리는게 좋아 ~
그럼 굳이 else로 depth가 늘어나지 않을듯!

kang-jisu reviewed (approved):
깔끔하게 잘 해준거같아~ 막 더 복잡한 설계를 더 해보라고 하긴 좀 애매해서 ..
시간이 남는다면 for문, if문 depth가 깊어지는 부분들을 한번 추출해서 메서드화시켜보는 연습을 해봐도 좋을거같넹 !!

YooJisu826 committed:
Update : orderMenu 메서드 간결화, if문 depth 줄이기

orderHistory 필드를 List<Menu>로 수정, Menu 클래스 내 필드명 변경,
payMoney 메서드 예외처리 추가, findMenu 메서드 추가

YooJisu826 commented:
코멘트 달아주신 부분 코드 수정해서 커밋했습니다!

구현에 집중해서 코드를 짜다보니 if문 depth 같은 부분은 생각하지 못 했던 것 같습니다! 다음부터는 이런 부분을 생각하면서 짜보겠습니다 감사합니다!
